### ISSUE TITLE
First Fiji launch has significant delay (due to Windows antivirus)

### ISSUE THREAD

**Issue by hinerm** [fiji/fiji#367]
(Edited after more testing)

On my Windows x64 system there is a significant delay during startup, but only the first time Fiji is launched in a particular context.
Scenarios where I can reproducibly cause long loading times:
* The first launch of the day, e.g. after the computer has been sleeping for several hours
* After restarting (but **not** after manually sleeping and immediately resuming)
* Duplicated `Fiji.app` (copy and paste `Fiji.app` and the first time running the copy will be slower)

The [stack trace](https://github.com/fiji/fiji/issues/367#issuecomment-2476766253) suggests this happens while looking for a custom backend logger. Here is _when_ the delay occurs varies slightly by how I'm launching Fiji:

- Standard mainline Fiji: after the splash screen pops, before the main UI loads.
- Mainline Fiji with Jaunch: after startup when the initial "Starting ImageJ" appears, before the text changes in appearance (which happens after the LAF is applied).
- Fiji future: between the the prompt window popping/disappearing, and the splash screen with "Launching Fiji". _Interestingly_, this splash screen also has an initial text display style that then changes after a second or two - but the "delay period" is _before_ the splash screen even shows.

The delay time is usually in the 5-10s neighborhood. When I started testing Fiji future it was regularly ~45s. However, although I can reproduce the delay consistently, I haven't found any correlation with its magnitude (and have not seen 45s delays since the first 2-3 days of testing).

**Comment by hinerm** [comment]
I'd like to better understand how this check is implemented and if we could make some changes w.r.t caching and/or running in parallel in the background..

**Comment by ctrueden** [comment]
Please post a thread dump of what's happening dueing those 45 sec. You can use `jstacktrace` or `jstack` from a separate Command Prompt (I forget the exact executable name).

**Comment by ctrueden** [comment]
Also: the update check already runs in a dedicated thread as a SciJava command. It should not block the appearance of the UI.

I wonder instead if Jaunch itself is getting hung up before net.imagej.Main starts? Or maybe a SciJava service is doing something network-related in its initialize methos -- if so, I would consider that a bug.

**Comment by hinerm** [comment]
>Tey also launching with --debug > ~/fflog.txt 2>&1

Nothing obvious here yet.

>Please post a thread dump 

This has been difficult to obtain since I only get one shot a day. Today I tried using `visualvm` to get the dump and attempting to connect to Fiji's pid during the startup *froze visualvm* until the splash screen popped.

I've been looking for ways to reproduce without needing to wait a day. Deleting all registry entries in `\HKEY_CURRENT_USER\Software\JavaSoft\Prefs` did not help.

**Comment by ctrueden** [comment]
I'm wondering if it's somehow I/O-related. You could try moving Fiji to a different physical medium, such as a USB stick.

Does `jstack` also freeze? Does the problem happen again if you reboot the computer? Or really only once per 24 hours?

**Comment by joshmoore** [comment]
🫨 

@hinerm: are you restarting between hand? I'm wondering if it's temp-directory related.

**Comment by hinerm** [comment]
>Does the problem happen again if you reboot the computer? Or really only once per 24 hours?
> are you restarting between hand?

@ctrueden oh! It does happen when restarting! But interestingly *not* when I just sleep the PC, which is what I usually do between days (and not a full restart).

**Comment by ctrueden** [comment]
@hinerm Can you try changing the `fiji.toml` to prepend the `app-launcher.jar` to the classpath?

```diff
--- a/home/curtis/code/fiji/fiji/config/jaunch/fiji.toml
+++ b/home/curtis/Apps/Fiji-Future.app/config/jaunch/fiji.toml
@@ -97,6 +97,7 @@ jvm.version-min = '8'
 
 jvm.classpath = [
     # Construct the classpath.
+    '${app-dir}/jars/app-launcher*.jar',
     '${app-dir}/jars/*.jar',
     '${app-dir}/jars/*/*.jar',
     '--plugins|${plugins}/*.jar'
```

It may be that because the system classpath is so long, and opening files so slow on Windows, that it will be able to find the main entry point much faster if the JAR containing it is first on the list. But even if this works, and results in the splash screen appearing immediately, it'll still be 45 seconds or more before the splash screen disappears, as all JARs are opened to obtain the `META-INF/json/org.scijava.plugin.Plugin` manifests for the SciJava service plugins.

@hinerm Have you had a chance to copy Fiji-Future.app onto a different storage medium (e.g. a USB stick over USB3 formatted with a not-FAT filesystem) to see if that makes any difference in startup performance?

**Comment by hinerm** [comment]
@ctrueden 

>Can you try changing the fiji.toml to prepend the app-launcher.jar to the classpath?

This didn't seem to change anything for me (delay happened before the splash screen popped)

>Have you had a chance to copy Fiji-Future.app onto a different storage medium

It didn't change anything, _but_ I did discover that I can reproduce the delay 100% consistently when I duplicate the `.app` directory. I updated the initial issue with my current knowledge.

**Comment by hinerm** [comment]
I'm not 100% sure if this is it, but I noticed that when I *build jaunch from source* and copy the launchers over, I get this notification (and a significant delay) the first time I launch:

![image](https://github.com/user-attachments/assets/c51925d5-1432-4d61-b5a5-045ac4fb87de)

I have smart app protection turned off in security settings and no idea where this scan is coming from at the moment.

Using the copy + paste trick to reproduce doesn't re-trigger this security warning notification, though. So I'm not sure if it's happening silently or if it may be a case of "this and something else". If I could figure out how to disable it I could verify, at least.

**Comment by ctrueden** [comment]
@hinerm This issue came up also at the recent Micro-Manager hackathon, with @nicost spending time working on how to sign Windows executables to minimize this nonsense. I asked Claude AI about it; here's what it said:

<blockquote>

This behavior occurs because Microsoft Defender performs a cloud-based security check on unknown executables through its Microsoft Defender SmartScreen feature. When you run a newly created/downloaded executable for the first time, Windows checks various security attributes:

1. Digital signature status
2. Reputation data from Microsoft's cloud security services
3. File hash and other security metrics

The main reasons this happens with your Kotlin Native executable are:

1. The executable is newly created and hasn't established a reputation in Microsoft's security database
2. It's likely unsigned (no code signing certificate)

Here are the main approaches to address this:

1. Code Signing: The most professional solution is to sign your executable with a valid code signing certificate. This establishes trust and reduces or eliminates the SmartScreen check.

```shell
# Example using signtool (Windows SDK)
signtool sign /tr http://timestamp.digicert.com /td sha256 /fd sha256 /a your_app.exe
```

2. Build Reputation:

* Over time, as more users run your signed executable without issues, it builds reputation in Microsoft's database
* This happens naturally through Microsoft's "Windows Defender SmartScreen" telemetry

3. File Properties: Ensure your executable has proper version information and metadata. While this won't eliminate the check, it can help build trust:

```gradle
// In your build.gradle.kts
windows {
    binaries {
        executable {
            windowsResources {
                versionInfo {
                    productName = "Your App Name"
                    companyName = "Your Company"
                    fileDescription = "Description"
                    fileVersion = "1.0.0.0"
                    productVersion = "1.0.0.0"
                }
            }
        }
    }
}
```

The most effective solution is code signing. While it has an associated cost (certificates typically cost a few hundred dollars annually), it provides several benefits:

1. Reduces or eliminates SmartScreen delays
2. Builds trust with users
3. Shows your executable hasn't been tampered with
5. Required for certain Windows features and Microsoft Store distribution

Note that completely disabling this security check isn't recommended as it's an important security feature. Instead, focus on establishing trust through proper signing and metadata.

</blockquote>

**Comment by hinerm** [comment]
OK, here is the Windows security setting that controls this scan. Turning it off removes the delay.

![image](https://github.com/user-attachments/assets/1ed1bf5c-c402-4ed9-b996-d7e17e295e82)

But since it turns itself on again periodically, it is better to go down and add an exclusion for your `Fiji.app`:
![image](https://github.com/user-attachments/assets/de323902-3201-4912-b743-6890c2dca8e7)

**Comment by hinerm** [comment]
Closing in favor of #377
